Adding L3DAS22 Task1 model to ESPNet-SE

popcornell opened the pull request:
As in the title, I am adding my re-implementation of the model we used to win the L3DAS22 speech enhancement challenge. 

Should I make two different files for MISO+BF+MISO and DenseUNet ? 
Also note that DenseUNet right now can not easily modified. Do you have any ideas on how we can make it more adaptable to e.g. different STFT window sizes ? 

**NOTE**:
WIP so don't run the tests for now as they will fail.

ftshijt commented:
We have fixed the CI with the latest master branch, can you merge that to fix the current CI issue? Many thanks

ftshijt commented:
Could you pull the latest master in espnet so as to fix the broken CI?

nateanl commented:
The formatting changes can be moved in a separate PR so that the main changes are more clearer.

Emrys365 commented:
> The formatting changes can be moved in a separate PR so that the main changes are more clearer.
> 

I thin they will be removed after applying the latest `black`.

popcornell commented:
Tests are failing because I am using Asteroid for some layers (especially filterbanks). 
Asteroid dependency can be removed, but asteroid_filterbanks is a light dependency and is very handy IMO.

sw005320 commented:
I see. Let me think. Like the other examples, we can include the installation of asteroid in the test (e.g., longformer)
https://github.com/espnet/espnet/pull/3992/files#diff-e327741dae7d1efd73c66ab134d50ac2c56187d814c443bb55464272b8b44113
https://github.com/espnet/espnet/pull/3992/files#diff-f8a2c3609ac8109e20d526231ec55513712deffde11895c1a93dd1c38355a077

popcornell commented:
Yeah, well actually asteroid can take a bit to install, because of PESQ basically and `mir_eval`. 
But asteroid_filterbanks is minimal and very handy for manipulating native complex torch. 

I am sure that these dependencies can be replaced by ESPNet-SE functions already, but I ll leave this in this state till INTERSPEECH deadline as the current architecture works in our experiments.

sw005320 commented:
I agree.

Emrys365 commented on espnet2/enh/separator/ineube_separator.py:
```diff
@@ -0,0 +1,147 @@
+        freeze_dnn1=False,
+    ) -> Tuple[List[Union[torch.Tensor, ComplexTensor]], torch.Tensor, OrderedDict]:
+
+        mix_stft = self.enc(input)
```
In my local trial, I encountered the following error:
```python
Traceback (most recent call last):
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "espnet-0.10.7a1/espnet2/bin/enh_train.py", line 23, in <module>
    main()
  File "espnet-0.10.7a1/espnet2/bin/enh_train.py", line 19, in main
    EnhancementTask.main(cmd=cmd)
  File "espnet-0.10.7a1/espnet2/tasks/abs_task.py", line 1019, in main
    cls.main_worker(args)
  File "espnet-0.10.7a1/espnet2/tasks/abs_task.py", line 1315, in main_worker
    cls.trainer.run(
  File "espnet-0.10.7a1/espnet2/train/trainer.py", line 286, in run
    all_steps_are_invalid = cls.train_one_epoch(
  File "espnet-0.10.7a1/espnet2/train/trainer.py", line 522, in train_one_epoch
    retval = model(**batch)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/espnet_model.py", line 160, in forward
    l, s, o = loss_wrapper(speech_ref, speech_pre, o)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/loss/wrappers/fixed_order.py", line 31, in forward
    loss += torch.mean(self.criterion(r, i)) / num_spk
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/loss/criterions/time_domain.py", line 165, in forward
    target_stft = stft_enc(target)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/enc_dec.py", line 177, in forward
    spec = multishape_conv1d(
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/scripting.py", line 37, in wrapper
    return fn(*args, **kwargs)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/enc_dec.py", line 210, in multishape_conv1d
    return F.conv1d(waveform.unsqueeze(1), filters, stride=stride, padding=padding)
RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.FloatTensor) should be the same
```
Is it as expected?

popcornell replied:
I am moving all the code to native ESPNet-SE now

Emrys365 commented on espnet2/enh/separator/ineube_separator.py:
```diff
@@ -0,0 +1,147 @@
+        tcn_repeats=4,
+        tcn_blocks=7,
+        tcn_channels=384,
+        activation=torch.nn.ELU,
```
It is better to use a string for choosing different activation functions.
Because in ESPnet, arguments are parsed directly from the yaml file. Strings are better supported than language-specific object.

popcornell replied:
I added an util for doing that https://github.com/popcornell/espnet/blob/177fc0d2eabbdbe531f5289437991603325a52fe/espnet2/torch_utils/get_layer_from_string.py#L5

Emrys365 commented on espnet2/enh/separator/ineube_separator.py:
```diff
@@ -0,0 +1,147 @@
+        output_from="dnn1",
+        n_chunks=3,
+        freeze_dnn1=False,
+    ) -> Tuple[List[Union[torch.Tensor, ComplexTensor]], torch.Tensor, OrderedDict]:
```
On second thought, I think it is better to move the arguments `output_from="dnn1"`, `n_chunks=3`, and `freeze_dnn1=False` to `__init__`, because it is messy to add model-specific argument in the current common [espnet_model](https://github.com/espnet/espnet/blob/master/espnet2/enh/espnet_model.py#L140).

In my recent trial, I find it convenient enough to move these arguments to model attributes, so that
* training of dnn1 or dnn2 can be configured directly in a config file
* it is more clear to understand the model behavior through the model config

---

What I did is as follows:
1. remove arguments `output_from`, `n_chunks`, and `freeze_dnn1`.
2. add attributes `train_dnn1`,  `train_dnn2`, and `n_chunks` in `__init__`.
3. add some code in `__init__` to freeze parameter accordingly:
   > ```python
   > if not self.train_dnn1:
   >    # freeze dnn1
   > elif not self.train_dnn1:
   >    # freeze dnn2
   > ```
4. modify the logic in `forward` to:
   > ```python
   > if self.train_dnn2:
   >    # same as output_from=="dnn2"
   > elif self.train_dnn1:
   >    # same as output_form=="dnn1"
   > else:
   >    # same as output_form=="beam"
   > ```

zqwang7 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,382 @@
+        )
+        self.post_blocks.append(last)
+
+    def forward(self, input):
```
what type of GPU are you using? this needs many memeory usage. This is just a comment, not a problem at all.

popcornell replied:
NVIDIA A100 40GB. Is it more memory intensive than using e.g. a torch Sequential wrapper ?

zqwang7 replied:
I meant the densenet concats features maps and uses many memory. There are some memory-efficient ways to reduce the memory consumption by re-computing some tensors during back-propagation.

popcornell replied:
like this for example ? https://github.com/gpleiss/efficient_densenet_pytorch/blob/master/models/densenet.py not really familiar with densenets

zqwang7 replied:
Yeah. it is what I was talking about. they used torch.utils.checkpoint to save training-time memory.

zqwang7 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,382 @@
+    ):
+        super(TCNDenseUNet, self).__init__()
+        self.n_spk = n_spk
+        self.in_channels = in_channels
```
"channel" has many meanings in speech processing. We may put a comment that here channel mean num_freqs.

zqwang7 commented on espnet2/enh/separator/ineube_separator.py:
```diff
@@ -0,0 +1,147 @@
+        )
+
+    def unfold(self, tf_rep, chunk_size):
+        bsz, freq, _ = tf_rep.shape
```
Maybe add some docstrings?

zqwang7 reviewed (approved):
I left some minor comments. I think the code looks good to me.

Emrys365 commented:
In addition, could you add some unit tests in [`test/espnet2/enh/test_espnet_model.py`](https://github.com/espnet/espnet/blob/master/test/espnet2/enh/test_espnet_model.py) and `test/espnet2/enh/separator/test_ineube_separator.py`?

> You could take existing unit tests under [`test/espnet2/enh/`](https://github.com/espnet/espnet/blob/master/test/espnet2/enh) as a reference.

Emrys365 commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -4,6 +4,8 @@
 import torch
 
 from espnet2.enh.loss.criterions.abs_loss import AbsEnhLoss
+import asteroid_filterbanks.transforms as af_transforms
+from asteroid_filterbanks import make_enc_dec
```
Could you sort the imports to follow the [PEP8 style](https://peps.python.org/pep-0008/#imports), i.e.:
```
<officiai python modules>

<third-party modules>

<espnet internal modules>
```

kamo-naoyuki commented:
Hi, are you still working for this PR? We are interested in this model.

popcornell commented:
Not lately, I am having an hard time finding time. 

However the model and the loss function are fine, it just needs the asteroid functions to be converted to ESPNet ones (e.g. STFT).

popcornell commented:
I removed asteroid dependencies. 
@neillu23 can you try running an experiment ? Tests are passing, model seems correct. I added extension to multiple speakers.

neillu23 commented:
> I removed asteroid dependencies. 
> 
> @neillu23 can you try running an experiment ? Tests are passing, model seems correct. I added extension to multiple speakers. 

Thanks to @popcornell! Let me run it on the L3DAS22 dataset.

popcornell commented:
@neillu23 do you have some results ?

neillu23 commented:
> @neillu23 do you have some results ?

I'm trying it out on the l3das22 dataset. Training is not over yet, but it looks good!
<img width="400" alt="截圖 2022-06-29 上午12 26 15" src="https://user-images.githubusercontent.com/62701324/176231697-61144631-f00d-4d62-af92-1a53b59e8455.png">

zican1998 commented:
> 

Why does the result of this network seem to have AGC capabilities? The amplitude of the enhanced speech remains the same

popcornell commented:
It is trained with variance normalized inputs and targets. Loss itself is scale invariant but that makes the output amplitude not fluctuating much.

zican1998 commented:
> 

![old](https://user-images.githubusercontent.com/105201243/176573729-0746f9e0-e871-4cfe-897b-1e9ea3de36c0.PNG)
![new](https://user-images.githubusercontent.com/105201243/176573748-fc1553e2-0c12-41fe-90b5-8de7d51ba537.PNG)
The first picture is the noisy speech, the second picture is the enhanced speech. The amplitude of the enhanced speech is very large and all scaled to the same amplitude. （Due to the large amplitude of the enhanced speech, I reduced the multiple when I tested it.）
Is this due to the scaling factor in the loss function? I feel confused.

nateanl commented:
> The amplitude of the enhanced speech is very large and all scaled to the same amplitude.

@zican1998  The scale of enhanced speech is uncontrolled if the loss function is si_snr. Even without the scaling factor, the amplitude of the enhanced speech could be very large.

zican1998 commented:
> The scale of enhanced speech is uncontrolled if the loss function is si_snr.
![image](https://user-images.githubusercontent.com/105201243/176581860-271b09f2-0c25-4da2-8ecf-293837f30eeb.png)
The loss function I use is MultiResL1SpecLoss written by the author. But why did the original fluctuating speech become consistent?

popcornell commented:
I am not sure I understood well but the loss function used is
scale-invariant with scaling factor same as SI-SDR. The scaling factor
however is applied to scale back the estimate to match the variance
normalized oracle clean speech. This leads to what you are observing.

On Wed, Jun 29, 2022, 10:48 PM zican1998 ***@***.***> wrote:

> The scale of enhanced speech is uncontrolled if the loss function is
> si_snr.
> [image: image]
> <https://user-images.githubusercontent.com/105201243/176581860-271b09f2-0c25-4da2-8ecf-293837f30eeb.png>
> The loss function I use is MultiResL1SpecLoss written by the author. But
> why did the original fluctuating speech become consistent?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/espnet/espnet/pull/3994#issuecomment-1170690264>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEO36ONZNSYLPMPTTCIFML3VRUDJBANCNFSM5MX4CAJA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Bob-Hu commented:
I made some attempts. If remove the scale factor, the magnitude of the result will not be pulled to the same level. Then this phenomenon seems to be related to the loss function. However, this can also happen if replace the loss function and use COS as the loss function. (And, under the same dataset, using COS as the loss function, under the DPCRN model, this won't happen). This phenomenon again seems to be related to ESPNET's model construction. It's really confusing.

nateanl commented:
> The loss function I use is MultiResL1SpecLoss written by the author. But why the original fluctuating speech becomes consistent?

I see. Sorry I was thinking the loss is si_snr when I looked at the tensorboard result. Yes the loss should be MultiResL1SpecLoss. And since it has a scaling factor, the magnitude is also dynamic similar to si_snr.

However, why the enhanced speech has AGC capability, that's an interesting question. @zqwang7 Do you know what's the potential reason?

Bob-Hu commented:
> It is also scale invariant so maybe that is the reason.
> […](#)
> On Thu, Jun 30, 2022, 9:19 AM Bob-Hu ***@***.***> wrote: @Bob-Hu <https://github.com/Bob-Hu> pardon my ignorance, what's COS loss function, cosine similarity ? Yes, It's cosine similarity — Reply to this email directly, view it on GitHub <[#3994 (comment)](https://github.com/espnet/espnet/pull/3994#issuecomment-1171210836)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEO36OMQAW4QVFDK2KFAQB3VRWNGNANCNFSM5MX4CAJA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Unlikely, because using cosine similarity under other models, there is no such phenomenon.

popcornell commented:
yes but the model you used is based on STFT masking.

Bob-Hu commented:
> yes but the model you used is based on STFT masking.

Maybe. Thank you very much for discussing with you.

popcornell commented:
@neillu23 sorry to ping you. Did your experiments have been successful ?

neillu23 commented:
> @neillu23 sorry to ping you. Did your experiments have been successful ?

@popcornell Yes, the results for DNN1 are quite good on the L3DAS22 recipe (with snr loss). If you think the results are close enough I can upload the model, or else we can fine-tune it with the multi-resolution L1 loss.

<img width="1024" alt="image" src="https://user-images.githubusercontent.com/62701324/178015901-b7c39b17-b384-4270-bfe4-b74895bf873e.png">

neillu23 commented:
I applied the default setting, so it would be DNN1.

Emrys365 commented:
This is great! Could you also update the PR? I will start a review after the code is ready.

neillu23 commented:
Thanks! @Emrys365 
@popcornell shall I make a PR to your branch for updating the results? Besides, I tried `output_form:dnn2` on CHiME4 but it shows "`constant_pad_nd(): argument 'input' (position 1) must be Tensor, not ComplexTensor`" at https://github.com/espnet/espnet/blob/77e36afdd3f069567dd33d4b5b997a26b634772b/espnet2/enh/separator/ineube_separator.py#L148

popcornell commented:
@neillu23 what torch version are you using ? I think you are using torch with no native complex support hence the error. 
I did not supported ComplexTensor in the MFMCWF code. Tests are running smoothly here and on my side so I cannot reproduce.

popcornell commented:
@Bob-Hu was great to discussing with you. Sorry I missed your last message. Did you find out if the masking is the cause ?

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+        upsample=False,
+        activation=torch.nn.ELU,
+    ):
+        super(Conv2DActNorm, self).__init__()
```
Could you replace this line with the Python3-style one?
```python
        super().__init__()
```

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+    """
+
+    def __init__(self, in_channels, num_freqs, out_channels, activation=torch.nn.ELU):
+        super(FreqWiseBlock, self).__init__()
```
ditto

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+        hid_chans=32,
+    ):
+
+        super(DenseBlock, self).__init__()
```
ditto

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+    def __init__(
+        self, in_chan, out_chan, ksz=3, stride=1, dilation=1, activation=torch.nn.ELU
+    ):
+        super(TCNResBlock, self).__init__()
```
ditto

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+        tcn_channels=384,
+        activation=torch.nn.ELU,
+    ):
+        super(TCNDenseUNet, self).__init__()
```
ditto

Emrys365 commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -261,3 +266,95 @@ def forward(self, ref, inf) -> torch.Tensor:
+    """
+
+    def __init__(self, window_sz=[512], hop_sz=None, eps=1e-8, time_domain_weight=0.5):
+        super(MultiResL1SpecLoss, self).__init__()
```
ditto

Emrys365 commented on espnet2/enh/layers/tcndenseunet.py:
```diff
@@ -0,0 +1,463 @@
+
+
+class FreqWiseBlock(torch.nn.Module):
+    """FreqWiseBlock, see iNeuBe paper.
```
Could you link to the paper here?

Emrys365 commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -261,3 +266,95 @@ def forward(self, ref, inf) -> torch.Tensor:
+    def __init__(self, window_sz=[512], hop_sz=None, eps=1e-8, time_domain_weight=0.5):
+        super(MultiResL1SpecLoss, self).__init__()
+
+        assert [x % 2 == 0 for x in window_sz]
```
I guess you wanted to write:
```python
        assert all([x % 2 == 0 for x in window_sz])
```

popcornell replied:
you are right

Emrys365 commented on espnet2/enh/separator/ineube_separator.py:
```diff
@@ -0,0 +1,147 @@
+class iNeuBe(AbsSeparator):
+    def __init__(
+        self,
+        n_spk=1,
```
Could you reflect the above comment?

Emrys365 commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -261,3 +266,95 @@ def forward(self, ref, inf) -> torch.Tensor:
+            weight for time domain loss.
+    """
+
+    def __init__(self, window_sz=[512], hop_sz=None, eps=1e-8, time_domain_weight=0.5):
```
Could you add the following two arguments like the others?
```
name=None, only_for_test=False
```

Emrys365 commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -261,3 +266,95 @@ def forward(self, ref, inf) -> torch.Tensor:
+
+        return stft.abs()
+
+    def forward(
+        self,
+        target: torch.Tensor,
+        estimate: torch.Tensor,
+    ):
```
```suggestion
    def forward(self, target: torch.Tensor, estimate: torch.Tensor):
```

Bob-Hu commented:
> @Bob-Hu was great to discussing with you. Sorry I missed your last message. Did you find out if the masking is the cause ?

Hi, if the output of the model is a mask instead of a signal spectrum, that phenomenon will not occur, so the mask is the root cause.

popcornell commented:
I added a check for torch >= 1.9. I think it is ready for review.

sw005320 commented:
The high-level design looks good to me.
Please leave the remaining remarks (normalization and magnitude loss, etc.) here or source codes for our future development.
I want to leave @Emrys365 for more detailed reviews.

popcornell commented:
I added it in the model docstring.

popcornell commented:
I think this is ready for review now.

Emrys365 reviewed (commented):
LGTM in general. Once the above comments are resolved, I can merge this PR.

popcornell commented on espnet2/enh/loss/criterions/time_domain.py:
```diff
@@ -261,3 +266,95 @@ def forward(self, ref, inf) -> torch.Tensor:
         return l1loss
+
+
+class MultiResL1SpecLoss(TimeDomainLoss):
```
done !

popcornell replied:
Can you rerun the tests ?

Emrys365 replied:
Sure.

popcornell commented:
I don't have any clue why tests are failing now.

popcornell commented:
Magically today all tests are passing. Should we merge @Emrys365  ?

sw005320 commented:
I merged it!
